### PR TITLE
[WFLY-9789] Upgrade to RESTEasy 3.5 (JAX-RS 2.1 upgrade)

### PR DESCRIPTION
**Pull request by asoldano**
This replaces https://github.com/wildfly/wildfly/pull/10843 

This brings JAX-RS 2.1 RESTEasy implementation in WildFly. Besides JSR 370 (JAX-RS 2.1), the RESTEasy component upgrade also pulls few other planned new features.

Few JIRA links:
https://issues.jboss.org/browse/WFLY-9789
https://issues.jboss.org/browse/WFLY-6459
https://issues.jboss.org/browse/WFLY-9786
https://issues.jboss.org/browse/WFLY-9787

Dev analysis:
https://github.com/wildfly/wildfly-proposals/pull/16
https://github.com/wildfly/wildfly-proposals/pull/18
https://github.com/wildfly/wildfly-proposals/pull/19
https://github.com/wildfly/wildfly-proposals/pull/24

### CODE REVIEW

File: feature-pack/src/main/resources/modules/system/layers/base/org/eclipse/yasson/main/module.xml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  ~ Lesser General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.5" name="org.eclipse.yasson">
```
Comment (jamezp) [inline feature-pack/src/main/resources/modules/system/layers/base/org/eclipse/yasson/main/module.xml]: This needs to be removed as it's defined in the `servlet-feature-pack`

File: feature-pack/pom.xml [hunk 2]
```diff
@@ -1373,6 +1411,16 @@
             </exclusions>
         </dependency>
 
+        <dependency>
+            <groupId>org.eclipse</groupId>
+            <artifactId>yasson</artifactId>
+        </dependency>
```
Comment (jamezp) [inline feature-pack/pom.xml]: This shouldn't be required because it's defined in the `servlet-feature-pack`.

**Review by jamezp — CHANGES_REQUESTED** [review]

**Review by jamezp — APPROVED** [review]
This looks good to me now. @kabir do we need a final release before we can merge this?

File: feature-pack/src/main/resources/modules/system/layers/base/com/github/fge/jackson-coreutils/main/module.xml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  ~ Lesser General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.5" name="com.github.fge.jackson-coreutils">
```
Comment (rsvoboda) [inline feature-pack/src/main/resources/modules/system/layers/base/com/github/fge/jackson-coreutils/main/module.xml]: urn:jboss:module:1.7

File: feature-pack/src/main/resources/modules/system/layers/base/com/github/fge/json-patch/main/module.xml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  ~ Lesser General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.5" name="com.github.fge.json-patch">
```
Comment (rsvoboda) [inline feature-pack/src/main/resources/modules/system/layers/base/com/github/fge/json-patch/main/module.xml]: urn:jboss:module:1.7

File: feature-pack/src/main/resources/modules/system/layers/base/org/jboss/resteasy/resteasy-jaxrs/main/module.xml [hunk 5]
```diff
@@ -46,5 +65,7 @@
         <module name="javax.servlet.api"/>
         <module name="org.jboss.resteasy.resteasy-validator-provider-11" optional="true" services="export" export="true"/>
         <module name="org.jboss.logging"/>
+        <module name="org.reactivestreams" export="true"/>
+        <module name="org.eclipse.microprofile.restclient" export="true"/>
```
Comment (rsvoboda) [inline feature-pack/src/main/resources/modules/system/layers/base/org/jboss/resteasy/resteasy-jaxrs/main/module.xml]: org.reactivestreams and org.eclipse.microprofile.restclient are public modules

 * is it OK to expose all all their classes ?
 * there is no analysis for these thing, is it going to be unsupported for now ?
I can imagine both can be appreciated in micro-service world / wf-swarm.

 * if they are unsupported, can be the modules switched off and enabled on request - e.g. system property?
we need probably think about such functionality so jboss-modules support `<conditions>` for `module` elements similar to `<artifact`.  @kabir is there plan for such thing ? There should be :) We need to be able to hide some stuff which is still under development, but a) is tight to something else and the relevant part is done b) we need feedback from early adopters of such feature

**Review by rsvoboda — CHANGES_REQUESTED** [review]
Biggest concern is how to deal with stuff like org.reactivestreams and org.eclipse.microprofile.restclient.

For WF 12 I think we could let it in if there is now way how to hide this stuff and enable it on customer request

For WF 13 we will need definitely means to hide stuff which is unsupported / wip

File: feature-pack/src/main/resources/modules/system/layers/base/org/jboss/resteasy/resteasy-jaxrs/main/module.xml (hunk 5 above)
Comment (asoldano) [reply feature-pack/src/main/resources/modules/system/layers/base/org/jboss/resteasy/resteasy-jaxrs/main/module.xml]: I definitely want to provide the analysis for those 2 things, just give me some time.
In the mean time, we can declare the module as privates if that really concerns you, or we can simply state that the features are tech preview in WFLY 12.

File: feature-pack/src/main/resources/modules/system/layers/base/com/github/fge/jackson-coreutils/main/module.xml (hunk 3 above)
Comment (asoldano) [reply feature-pack/src/main/resources/modules/system/layers/base/com/github/fge/jackson-coreutils/main/module.xml]: Not strictly needed, but I can change this

File: feature-pack/src/main/resources/modules/system/layers/base/com/github/fge/json-patch/main/module.xml (hunk 4 above)
Comment (asoldano) [reply feature-pack/src/main/resources/modules/system/layers/base/com/github/fge/json-patch/main/module.xml]: Not strictly needed, but I can change this

**Comment by asoldano** [comment]
@rsvoboda , the resteasy-jaxrs module does export the microprofile rest client stuff anymore, so that if we want to mark that module as unsupported for now, we can do that. In any case, I've provided initial dev analysis.
The org.reactivestreams module contains a single jar; see http://www.reactive-streams.org/reactive-streams-1.0.1-javadoc/ , the contents are 4 interfaces only.

File: feature-pack/src/main/resources/modules/system/layers/base/org/jboss/resteasy/resteasy-jaxrs/main/module.xml (hunk 5 above)
Comment (rsvoboda) [reply feature-pack/src/main/resources/modules/system/layers/base/org/jboss/resteasy/resteasy-jaxrs/main/module.xml]: +1 for private module

**Review by rsvoboda — APPROVED** [review]

**Pull request merged**